icu4j/main/classes/core/src/com/ibm/icu/number/NumberFormatterImpl.java

```diff
@@ -353,14 +342,43 @@ private static MicroPropsGenerator macrosToMicroGenerator(MacroProps macros, Mic
                 // Lazily create PluralRules
                 rules = PluralRules.forLocale(macros.loc);
             }
-            chain = LongNameHandler
-                    .forMeasureUnit(macros.loc, macros.unit, macros.perUnit, unitWidth, rules, chain);
+            PluralRules pluralRules = macros.rules != null ?
+                    macros.rules :
+                    PluralRules.forLocale(macros.loc);
+
+            if (macros.usage != null) {
+                chain = LongNameMultiplexer.forMeasureUnits(
+                        macros.loc,
+                        micros.outputUnit.splitToSingleUnits(),
```

Comment:
Look again:

                LongNameMultiplexer::forMeasureUnits(
                    macros.locale, *fUsagePrefsHandler->getOutputUnits(), unitWidth,
                    resolvePluralRules(macros.rules, macros.locale, status), chain, status),
                status);

    /**
     * Returns the list of possible output units, i.e. the full set of
     * preferences, for the localized, usage-specific unit preferences.
     *
     * The returned pointer should be valid for the lifetime of the
     * UsagePrefsHandler instance.
     */

You're not taking the list of output units related to the usage, you're taking only one outputUnit and splitting it to single units.

Reply:
You need to call fUnitsRouter.getOutputUnits. fUnitsRouter is in the UsagePrefsHandler class. UsagePrefsHandler needs a getOutputUnits() method.

Reply:
Younies already did this. Marking as resolved.